shared/js/background/declarative-net-request.js

```diff
@@ -0,0 +1,209 @@
+        }
+    }
+
+    // Otherwise, it is necessary to update the declarativeNetRequest rules and
+    // settings again.
+
+    let addRules
+    let lookup
+
+    // Tracker blocking.
+    if (configName === 'tds') {
+        const {
+            ruleset, trackerDomainByRuleId
+        } = await generateTrackerBlockingRuleset(
+            configValue,
+            chrome.declarativeNetRequest.isRegexSupported,
+            ruleIdStart + 1
+        )
+        addRules = ruleset
+        lookup = trackerDomainByRuleId
+    // Extension configuration.
+    } else if (configName === 'config') {
+        const {
+            ruleset, matchDetailsByRuleId
+        } = await generateExtensionConfigurationRuleset(
+            configValue,
+            chrome.declarativeNetRequest.isRegexSupported,
+            ruleIdStart + 1
+        )
+        addRules = ruleset
+        lookup = matchDetailsByRuleId
+    }
+
+    if (!addRules) {
+        console.error(
+            'No declarativeNetRequest rules generated for configuration: ',
+            configName, '(Etag: ', etag, ')'
+        )
+        return
+    }
+
+    // Add the new etag rule.
+    addRules.push(generateEtagRule(etagRuleId, etag))
+
+    if (addRules.length > maxNumberOfRules) {
+        console.error(
+            'Too many declarativeNetRequest rules generated for configuration: ',
+            configName,
+            '(Etag: ', etag, ', Rules generated: ', addRules.length, ')'
+        )
+        return
+    }
+
+    // Ensure any existing rules for the configuration are removed.
+    const removeRuleIds = []
+    for (let i = ruleIdStart; i <= ruleIdEnd; i++) {
+        removeRuleIds.push(i)
+    }
+
+    // Install the updated rules and then update the setting entry.
+    await chrome.declarativeNetRequest.updateDynamicRules({
+        removeRuleIds, addRules
+    })
+    settings.updateSetting(settingName, { etag, lookup, extensionVersion })
+}
+
+/**
+ * @typedef {object} getMatchDetailsTrackerBlockingResult
+ * @property {string} type
+ *   The match type 'trackerBlocking'.
+ * @property {string[]} possibleTrackerDomains
+ *   The list of possible tracking domains associated with this match.
+ */
+
+/**
+ * @typedef {object} getMatchDetailsExtensionConfigurationResult
+ * @property {string} type
+ *   The match type, for example 'trackerAllowlist'.
+ * @property {string} domain
+ *   The matching tracker domain.
+ * @property {string} reason
+ *   The reason for the match.
+ */
+
+/**
+ * @typedef {object} getMatchDetailsUnknownResult
+ * @property {string} type
+ *   The match type 'unknown'.
+ */
+
+/**
+ * Find the match details (if any) associated with the given
+ * declarativeNetRequest rule ID.
+ * @param {number} ruleId
+ * @return {Promise<getMatchDetailsUnknownResult |
+ *                  getMatchDetailsExtensionConfigurationResult |
+ *                  getMatchDetailsTrackerBlockingResult>}
+ */
+export async function getMatchDetails (ruleId) {
```

Comment:
Is this used anywhere currently?  Even if its not, it would be good to have some test coverage for it if its included in this PR.

Reply:
No, this isn't used yet. Good point about adding some tests though. Done!